src/python/pants/backend/project_info/rules/source_file_validator.py

```diff
@@ -0,0 +1,249 @@
+# coding=utf-8
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import re
+
+from future.utils import text_type
+
+from pants.engine.console import Console
+from pants.engine.fs import Digest, FilesContent
+from pants.engine.legacy.graph import HydratedTargets
+from pants.engine.objects import Collection
+from pants.engine.rules import rule, optionable_rule, console_rule
+from pants.engine.selectors import Get
+from pants.rules.core.exceptions import GracefulTerminationException
+from pants.subsystem.subsystem import Subsystem
+from pants.util.objects import datatype, enum
+
+
+class DetailLevel(enum(['none', 'summary', 'nonmatching', 'all'])):
+  """How much detail about validation to emit to the console.
+
+  none: Emit nothing.
+  summary: Emit a summary only.
+  nonmatched: Emit details for source files that failed to match at least one required pattern.
+  all: Emit details for all source files.
+  """
+  pass
+
+
+class SourceFileValidation(Subsystem):
```

Comment:
I think having it as a continuum makes some sense to me, depending on the degree of logic sharing. Only rules can do things like yield to the engine, so if that's needed it would tend towards making a v2 ruleset, but if rules are more just windows into existing subsystem logic (which they would be at first for existing code), it would tend to have more in the subsystem. I think for things that would normally be helper methods on `Subsystem` instances, it's probably better to move those into `@rule`s, and in general keep subsystems as a bag of options. Some existing mildly-complex functionality such as `CompilerOptionSetsMixin` could stay on subsystems too, though.

Reply:
Yea. I expect that (as with testing), we're going to want to factor linting as a `@console_rule` that consumes an `@union LintResult` (or something). Unlike with tests (which generally have one test runner), per-language linting will likely have an additional level of composition because one target type (ie `PythonLibrary`) might have many linters that need composition.

Reply:
Here might be a good example of taking logic similar to that implemented in this subsystem, but splitting it out into v2 rules: #7419. The only reason that PR didn't also create an `@console_rule` was because I wanted to make sure it ran when any v1 `gen` task did, and there's currently no way to expose a v2 rule directly as a v1 task except by doing what I did there and wrapping it in a v1 task, which seems good for now.